File: android-app/app/src/main/java/com/sherlock/android/service/SiteChecker.kt
package com.sherlock.android.service

import com.sherlock.android.model.CheckResult
import com.sherlock.android.model.ResultStatus
import com.sherlock.android.model.SiteInfo
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

class SiteChecker {

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .followRedirects(false)
        .build()

    private val clientFollowRedirects = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .followRedirects(true)
        .build()

    fun check(site: SiteInfo, username: String): CheckResult {
        if (site.regexCheck != null) {
            val regex = Regex(site.regexCheck)
            if (!regex.containsMatchIn(username)) {
                return CheckResult(
                    siteName = site.name,
                    url = site.url.replace("{}", username),
                    status = ResultStatus.NOT_FOUND,
                    errorMessage = "Username format not valid for this site",
                    isDanishDating = site.isDanishDating
                )
            }
        }

        val targetUrl = site.url.replace("{}", username)

        return try {
            when (site.errorType) {
                "status_code" -> checkStatusCode(site, targetUrl)
                "message" -> checkMessage(site, targetUrl)
                "response_url" -> checkResponseUrl(site, targetUrl)
                else -> CheckResult(site.name, targetUrl, ResultStatus.ERROR, "Unknown errorType", site.isDanishDating)
            }
        } catch (e: Exception) {
            CheckResult(site.name, targetUrl, ResultStatus.ERROR, e.message, site.isDanishDating)
        }
    }

    private fun checkStatusCode(site: SiteInfo, url: String): CheckResult {
        val request = Request.Builder().url(url).get().build()
        client.newCall(request).execute().use { response ->
            val found = response.isSuccessful
            return CheckResult(
                siteName = site.name,
                url = url,
                status = if (found) ResultStatus.FOUND else ResultStatus.NOT_FOUND,
                isDanishDating = site.isDanishDating
            )
        }
    }

    private fun checkMessage(site: SiteInfo, url: String): CheckResult {
        if (site.errorMsgs.isEmpty()) {
            return CheckResult(site.name, url, ResultStatus.ERROR, "No errorMsg defined", site.isDanishDating)
        }
        val request = Request.Builder().url(url).get().build()
        clientFollowRedirects.newCall(request).execute().use { response ->
            val body = response.body?.string() ?: ""
            val notFound = site.errorMsgs.any { msg -> body.contains(msg) }
            return CheckResult(
                siteName = site.name,
                url = url,
                status = if (notFound) ResultStatus.NOT_FOUND else ResultStatus.FOUND,
                isDanishDating = site.isDanishDating
            )
        }
    }

    // response_url sites return HTTP 200 for existing users but redirect (3xx)
    // to an error/login page when the user doesn't exist.
    private fun checkResponseUrl(site: SiteInfo, url: String): CheckResult {
        val request = Request.Builder().url(url).get().build()
        client.newCall(request).execute().use { response ->
            val found = response.isSuccessful
            return CheckResult(
                siteName = site.name,
                url = url,
                status = if (found) ResultStatus.FOUND else ResultStatus.NOT_FOUND,
                isDanishDating = site.isDanishDating
            )
        }
    }
}

File: mobile-app/lib/services/site_checker.dart
import 'dart:async';
import 'package:http/http.dart' as http;
import '../models/site_info.dart';
import '../models/check_result.dart';

class SiteChecker {
  static const _ua =
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:129.0) Gecko/20100101 Firefox/129.0';
  static const _timeout = Duration(seconds: 15);

  final http.Client _client;

  SiteChecker() : _client = http.Client();

  Future<CheckResult> check(SiteInfo site, String username) async {
    if (site.regexCheck != null) {
      final ok = RegExp(site.regexCheck!).hasMatch(username);
      if (!ok) {
        return CheckResult(
          siteName: site.name,
          url: site.url.replaceAll('{}', username),
          status: ResultStatus.notFound,
          errorMessage: 'Username format invalid for this site',
        );
      }
    }

    final targetUrl = site.url.replaceAll('{}', username);
    final probeUrl =
        (site.urlProbe ?? site.url).replaceAll('{}', username);

    final headers = <String, String>{
      'User-Agent': _ua,
      if (site.headers != null) ...site.headers!,
    };

    try {
      switch (site.errorType) {
        case 'status_code':
          return await _byStatusCode(site, targetUrl, probeUrl, headers);
        case 'message':
          return await _byMessage(site, targetUrl, probeUrl, headers);
        case 'response_url':
          return await _byResponseUrl(site, targetUrl, probeUrl, headers);
        default:
          return CheckResult(
            siteName: site.name,
            url: targetUrl,
            status: ResultStatus.error,
            errorMessage: 'Unknown errorType: ${site.errorType}',
          );
      }
    } on TimeoutException {
      return CheckResult(
        siteName: site.name,
        url: targetUrl,
        status: ResultStatus.error,
        errorMessage: 'Timeout',
      );
    } catch (e) {
      return CheckResult(
        siteName: site.name,
        url: targetUrl,
        status: ResultStatus.error,
        errorMessage: e.toString(),
      );
    }
  }

  Future<CheckResult> _byStatusCode(
    SiteInfo site,
    String targetUrl,
    String probeUrl,
    Map<String, String> headers,
  ) async {
    final res = await _client
        .head(Uri.parse(probeUrl), headers: headers)
        .timeout(_timeout);
    return CheckResult(
      siteName: site.name,
      url: targetUrl,
      status: _statusOk(res.statusCode)
          ? ResultStatus.found
          : ResultStatus.notFound,
    );
  }

  Future<CheckResult> _byMessage(
    SiteInfo site,
    String targetUrl,
    String probeUrl,
    Map<String, String> headers,
  ) async {
    final res = await _client
        .get(Uri.parse(probeUrl), headers: headers)
        .timeout(_timeout);
    final body = res.body;

    bool hasErrorMsg = false;
    final msg = site.errorMsg;
    if (msg is String) {
      hasErrorMsg = body.contains(msg);
    } else if (msg is List) {
      hasErrorMsg = msg.any((m) => body.contains(m.toString()));
    }

    return CheckResult(
      siteName: site.name,
      url: targetUrl,
      status: hasErrorMsg ? ResultStatus.notFound : ResultStatus.found,
    );
  }

  Future<CheckResult> _byResponseUrl(
    SiteInfo site,
    String targetUrl,
    String probeUrl,
    Map<String, String> headers,
  ) async {
    final res = await _client
        .get(Uri.parse(probeUrl), headers: headers)
        .timeout(_timeout);
    return CheckResult(
      siteName: site.name,
      url: targetUrl,
      status: _statusOk(res.statusCode)
          ? ResultStatus.found
          : ResultStatus.notFound,
    );
  }

  bool _statusOk(int code) => code >= 200 && code < 300;

  void dispose() => _client.close();
}
